examples/hybrid/staggered_nonhydrostatic_model.jl

```diff
@@ -273,46 +287,52 @@ end
 additional_tendency!(Yₜ, Y, p, t) = nothing
 
 function Wfact!(W, Y, p, dtγ, t)
-    (; flags, dtγ_ref, ∂ᶜρₜ∂ᶠ𝕄, ∂ᶜ𝔼ₜ∂ᶠ𝕄, ∂ᶠ𝕄ₜ∂ᶜ𝔼, ∂ᶠ𝕄ₜ∂ᶜρ, ∂ᶠ𝕄ₜ∂ᶠ𝕄) = W
+    (; ∂Yₜ∂Y, ∂R∂Y, transform, flags) = W
     ᶜρ = Y.c.ρ
     ᶜuₕ = Y.c.uₕ
     ᶠw = Y.f.w
-    (; ᶜK, ᶜΦ, ᶜp, ∂ᶜK∂ᶠw_data, ᶠupwind_product) = p
-
-    dtγ_ref[] = dtγ
-
-    # If we let ᶠw_data = ᶠw.components.data.:1 and ᶠw_unit = one.(ᶠw), then
-    # ᶠw == ᶠw_data .* ᶠw_unit. The Jacobian blocks involve ᶠw_data, not ᶠw.
-    ᶠw_data = ᶠw.components.data.:1
-
-    # If ∂(ᶜarg)/∂(ᶠw_data) = 0, then
-    # ∂(ᶠupwind_product(ᶠw, ᶜarg))/∂(ᶠw_data) =
-    #     ᶠupwind_product(ᶠw + εw, arg) / to_scalar(ᶠw + εw).
-    # The εw is only necessary in case w = 0.
-    εw = Ref(Geometry.Covariant3Vector(eps(FT)))
-    to_scalar(vector) = vector.u₃
-
-    # ᶜinterp(ᶠw) =
-    #     ᶜinterp(ᶠw)_data * ᶜinterp(ᶠw)_unit =
-    #     ᶜinterp(ᶠw_data) * ᶜinterp(ᶠw)_unit
-    # norm_sqr(ᶜinterp(ᶠw)) =
-    #     norm_sqr(ᶜinterp(ᶠw_data) * ᶜinterp(ᶠw)_unit) =
-    #     ᶜinterp(ᶠw_data)^2 * norm_sqr(ᶜinterp(ᶠw)_unit)
+    (; ᶜK, ᶜΦ, ᶜp, ∂ᶜK∂ᶠw) = p
+    (; ᶠupwind_product, ᶠupwind_product_matrix, ᶠno_flux_row) = p
+
+    ᶜρ_name = @name(c.ρ)
+    ᶜ𝔼_name = if :ρθ in propertynames(Y.c)
+        @name(c.ρθ)
+    elseif :ρe in propertynames(Y.c)
+        @name(c.ρe)
+    elseif :ρe_int in propertynames(Y.c)
+        @name(c.ρe_int)
+    end
+    ᶠ𝕄_name = @name(f.w)
+    ∂ᶜρₜ∂ᶠ𝕄 = ∂Yₜ∂Y[ᶜρ_name, ᶠ𝕄_name]
+    ∂ᶜ𝔼ₜ∂ᶠ𝕄 = ∂Yₜ∂Y[ᶜ𝔼_name, ᶠ𝕄_name]
+    ∂ᶠ𝕄ₜ∂ᶜρ = ∂Yₜ∂Y[ᶠ𝕄_name, ᶜρ_name]
+    ∂ᶠ𝕄ₜ∂ᶜ𝔼 = ∂Yₜ∂Y[ᶠ𝕄_name, ᶜ𝔼_name]
+    ∂ᶠ𝕄ₜ∂ᶠ𝕄 = ∂Yₜ∂Y[ᶠ𝕄_name, ᶠ𝕄_name]
+
+    ᶠgⁱʲ = Fields.local_geometry_field(ᶠw).gⁱʲ
+    g³³(gⁱʲ) = Geometry.AxisTensor(
+        (Geometry.Contravariant3Axis(), Geometry.Contravariant3Axis()),
+        Geometry.components(gⁱʲ)[end],
+    )
+
+    # If ∂(ᶜχ)/∂(ᶠw) = 0, then
+    # ∂(ᶠupwind_product(ᶠw, ᶜχ))/∂(ᶠw) =
+    #     ∂(ᶠupwind_product(ᶠw, ᶜχ))/∂(CT3(ᶠw)) * ∂(CT3(ᶠw))/∂(ᶠw) =
+    #     vec_data(ᶠupwind_product(ᶠw + εw, ᶜχ)) / vec_data(CT3(ᶠw + εw)) * ᶠg³³
+    # The ε is only necessary when w = 0. Since ᶠupwind_product is undefined at
+    # the boundaries, we also need to wrap it in a call to ᶠno_flux.
+    vec_data(vector) = vector[1]
```

Comment:
Can we document what `vec_data` is for?

Reply:
I've added some more details to this comment about `vec_data` and `εw`.